subsys/logging/log_backend_swo.c

```diff
@@ -36,18 +36,25 @@
 #if CONFIG_LOG_BACKEND_SWO_FREQ_HZ == 0
 #define SWO_FREQ_DIV  1
 #else
-#if DT_NODE_HAS_PROP(DT_PATH(cpus, cpu_0), clock_frequency)
-#define CPU_FREQ DT_PROP(DT_PATH(cpus, cpu_0), clock_frequency)
+
+/* Set reference frequency which can be custom or cpu frequency. */
+#if DT_NODE_HAS_PROP(DT_PATH(cpus, cpu_0), swo_ref_frequency)
+#define SWO_REF_FREQ DT_PROP(DT_PATH(cpus, cpu_0), swo_ref_frequency)
+#elif DT_NODE_HAS_PROP(DT_PATH(cpus, cpu_0), clock_frequency)
+#define SWO_REF_FREQ DT_PROP(DT_PATH(cpus, cpu_0), clock_frequency)
 #else
 #error "Missing DT 'clock-frequency' property on cpu@0 node"
 #endif
-#define SWO_FREQ (CPU_FREQ + (CONFIG_LOG_BACKEND_SWO_FREQ_HZ / 2))
-#define SWO_FREQ_DIV  (SWO_FREQ / CONFIG_LOG_BACKEND_SWO_FREQ_HZ)
+
+#define SWO_FREQ_DIV \
+	(SWO_REF_FREQ / CONFIG_LOG_BACKEND_SWO_FREQ_HZ)
```

Comment:
maybe there should be an error when target frequency cannot be matched?

Reply:
That would be an option, however a legal output frequency can be a rational number. Let's say - in the example above - user wants to set up SWO trace with divider equal to 3, to achieve output frequency of 10 MHz / 3 = 3.3333 MHz.

Reply:
added compilation error when target frequency cannot be achieved. Expecting that there will be up to 1 hz difference between reference and div * target to accommodate for fractional number.

Reply:
This will not work if we divide by a different number, e.g. 7. For other combinations of dividend and divisor the error can be larger. It's safer to round `SWO_FREQ_DIV` to the nearest integer using the formula I proposed.

If you are worried that the formula is not readable, maybe we should create a macro that would round to the nearest integer and put it in utils?

Reply:
Used your formula, can you revisit?